File: backend/src/main/java/com/project2/global/security/Rq.java
package com.project2.global.security;

import java.util.List;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import com.project2.domain.member.entity.Member;
import com.project2.domain.member.service.MemberService;
import com.project2.global.exception.ServiceException;

import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;

// Request, Response, Session, Cookie, Header
@Component
@RequiredArgsConstructor
public class Rq {

	private final HttpServletRequest request;
	private final HttpServletResponse response;
	private final MemberService memberService;

	@Value("${custom.url.domain}")
	private String siteDomain;

	public void setLogin(Member actor) {

		// 유저 정보 생성
		UserDetails user = new SecurityUser(actor.getId(), actor.getEmail(),
			List.of(new SimpleGrantedAuthority("ROLE_USER")));

		// 인증 정보 저장소
		SecurityContextHolder.getContext().setAuthentication(
			new UsernamePasswordAuthenticationToken(user, null, user.getAuthorities())
		);
	}

	public Member getActor() {
		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();

		if (authentication == null) {
			throw new ServiceException("401-2", "로그인이 필요합니다.");
		}

		Object principal = authentication.getPrincipal();

		if (!(principal instanceof SecurityUser)) {
			throw new ServiceException("401-3", "잘못된 인증 정보입니다");
		}

		SecurityUser user = (SecurityUser)principal;

		return Member.builder()
			.id(user.getId())
			.email(user.getUsername())
			.build();
	}

	public String getHeader(String name) {
		return request.getHeader(name);
	}

	public String getValueFromCookie(String name) {
		Cookie[] cookies = request.getCookies();

		if (cookies == null) {
			return null;
		}

		for (Cookie cookie : cookies) {
			if (cookie.getName().equals(name)) {
				return cookie.getValue();
			}
		}

		return null;
	}

	public void setHeader(String name, String value) {
		response.setHeader(name, value);
	}

	public void addCookie(String name, String value, boolean isHttpOnly) {
		Cookie accsessTokenCookie = new Cookie(name, value);

		accsessTokenCookie.setDomain(siteDomain);
		accsessTokenCookie.setPath("/");
		accsessTokenCookie.setHttpOnly(isHttpOnly);
		accsessTokenCookie.setSecure(true);
		accsessTokenCookie.setAttribute("SameSite", "Strict");

		response.addCookie(accsessTokenCookie);
	}

	public Member getRealActor(Member actor) {
		return memberService.findById(actor.getId()).get();
	}

	public void removeCookie(String name) {
		// 원칙적으로 쿠키를 서버에서 삭제하는 것은 불가능.

		Cookie cookie = new Cookie(name, null);
		cookie.setDomain(siteDomain);
		cookie.setPath("/");
		cookie.setHttpOnly(true);
		cookie.setSecure(true);
		cookie.setAttribute("SameSite", "Strict");
		cookie.setMaxAge(0);

		response.addCookie(cookie);
	}
}

File: backend/src/main/kotlin/com/project2/global/security/Rq.kt
package com.project2.global.security

import com.project2.domain.member.entity.Member
import com.project2.domain.member.service.MemberService
import com.project2.global.exception.ServiceException
import jakarta.servlet.http.Cookie
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Component

@Component
class Rq(
        private val request: HttpServletRequest,
        private val response: HttpServletResponse,
        private val memberService: MemberService,
        @Value("\${custom.url.domain}") private val siteDomain: String
) {

    fun setLogin(actor: Member) {
        val user: UserDetails = SecurityUser(
                actor.id!!,
                actor.email,
                listOf(SimpleGrantedAuthority("ROLE_USER"))
        )
        SecurityContextHolder.getContext().authentication =
                UsernamePasswordAuthenticationToken(user, null, user.authorities)
    }

    fun getActor(): Member {
        val authentication: Authentication = SecurityContextHolder.getContext().authentication
                ?: throw ServiceException("401-2", "로그인이 필요합니다.")

        val principal = authentication.principal
        if (principal !is SecurityUser) {
            throw ServiceException("401-3", "잘못된 인증 정보입니다")
        }

        return Member.builder()
                .id(principal.id)
                .email(principal.username)
                .build()
    }

    fun getHeader(name: String): String? {
        return request.getHeader(name)
    }

    fun getValueFromCookie(name: String): String? {
        val cookies = request.cookies ?: return null
        return cookies.firstOrNull { it.name == name }?.value
    }

    fun setHeader(name: String, value: String) {
        response.setHeader(name, value)
    }

    fun addCookie(name: String, value: String, isHttpOnly: Boolean) {
        val accessTokenCookie = Cookie(name, value).apply {
            domain = siteDomain
            path = "/"
            this.isHttpOnly = isHttpOnly
            secure = true
            setAttribute("SameSite", "Strict")
        }
        response.addCookie(accessTokenCookie)
    }

    fun getRealActor(actor: Member): Member {
        return memberService.findById(actor.id!!).orElseThrow()
    }

    fun removeCookie(name: String) {
        val cookie = Cookie(name, null).apply {
            domain = siteDomain
            path = "/"
            isHttpOnly = true
            secure = true
            setAttribute("SameSite", "Strict")
            maxAge = 0
        }
        response.addCookie(cookie)
    }
}
